pushtx/broadcaster.go

```diff
@@ -130,31 +137,51 @@ func (b *Broadcaster) broadcastHandler(sub *blockntfns.Subscription) {
 
 		// A new block notification has arrived, so we'll rebroadcast
 		// all of our pending transactions.
-		case block, ok := <-sub.Notifications:
+		case _, ok := <-sub.Notifications:
 			if !ok {
 				log.Warn("Unable to rebroadcast transactions: " +
 					"block subscription was canceled")
 				continue
 			}
 
-			blockHeader := block.Header()
-
-			if len(b.transactions) != 0 {
-				log.Debugf("Re-broadcasting transaction at "+
-					"height=%v, hash=%v", block.Height(),
-					blockHeader.BlockHash())
+			select {
+			// If the rebroadcast semaphore is available, start a
+			// new goroutine to exectue a rebroadcast.
+			case <-rebroadcastSem:
+			default:
+				log.Debugf("Existing rebroadcast still in " +
+					"progress")
+				continue
 			}
 
-			b.rebroadcast()
+			b.wg.Add(1)
+			go func() {
+				defer b.wg.Done()
+
+				b.rebroadcast()
+				rebroadcastSem <- struct{}{}
+			}()
 
 		// Between blocks, we'll also try to attempt additional
 		// re-broadcasts to ensure a timely confirmation.
 		case <-reBroadcastTicker.C:
-			if len(b.transactions) == 0 {
+			select {
+			// If the rebroadcast semaphore is available, start a
+			// new goroutine to exectue a rebroadcast.
+			case <-rebroadcastSem:
+			default:
+				log.Debugf("Existing rebroadcast still in " +
```

Comment:
Nit: this might be a little spammy, demote to trace?

Reply:
This should only happen if a rebroadcast take more than 1min || block time. Should be fine?